pytorch_lightning/core/lightning.py

```diff
@@ -91,11 +91,42 @@ def forward(self, x):
         if self.trainer.proc_rank == 0:
             print(*args, **kwargs)
 
+    def __call__(self, *data, **kwargs):
+        r"""
+        Automatically moves data to correct device if possible, then call torch.nn.Module.__call__
+        Lightning will warn you if it automatically moves any data
+
+        Args:
+            *data: Any positional arguments for torch.nn.Module.__call__. These are typically input data
+            **kwargs: Any keyword arguments for torch.nn.Module.__call__
+
+        Example:
+
+            .. code-block:: python
+
+                model = model.cuda(0)
+                model.prepare_data()
+                loader = model.train_dataloader()
+                for x, y in loader:
+                    output = model(x)  # Lightning will automove data here and warn you of it
+
+        """
+        devices = [p.device for p in self.parameters()]
```

Comment:
I'm not sure, if looping always over all the params is a good idea. Can we maybe cache the devices somehow?

Reply:
@mcarilli are we missing a simpler/cleaner way of doing this?

```python

x = x.cpu()
model.cuda()

# this works in lightning
out = model(x)

# out is cuda tensor now
```

Reply:
It looks like you're only using device[0], so why are you collecting them all?  Also if the model params do reside on multiple devices, it's hard to predict which device the user actually wants the data to reside on.

Reply:
@HenryJia

Reply:
So, I only apply automatic data transfer if we are dealing with the simple case of the model residing on one device, as trying to auto transfer data when the model is spread across multiple devices is very non-trivial and is heavily dependent on model structure